File: Source/Samples/Jacobi.Vst.Samples.Delay.VB/PluginPersistence.vb
Imports System.Diagnostics
Imports System.IO
Imports System.Text
Imports Jacobi.Vst.Core
Imports Jacobi.Vst.Framework
Imports Jacobi.Vst.Framework.Plugin
Imports Jacobi.Vst.Framework.Plugin.IO

Friend Class PluginPersistence
    Inherits VstPluginPersistenceBase

    Private _plugin As FxTestPlugin

    Public Sub New(ByVal plugin As FxTestPlugin)
        _plugin = plugin
    End Sub

    Protected Overrides Function CreateProgramReader(ByVal input As Stream) As VstProgramReaderBase
        Return New DelayProgramReader(_plugin, input, Encoding)
    End Function

    Private Class DelayProgramReader
        Inherits VstProgramReaderBase

        Private _plugin As FxTestPlugin

        Public Sub New(ByVal plugin As FxTestPlugin, ByVal input As Stream, ByVal encoding As Encoding)
            MyBase.New(input, encoding)

            _plugin = plugin
        End Sub

        Protected Overrides Function CreateProgram() As Framework.VstProgram
            Dim program As VstProgram = New VstProgram(_plugin.ParameterFactory.Categories)

            _plugin.ParameterFactory.CreateParameters(program.Parameters)

            Return program
        End Function

    End Class
End Class

File: Source.Core/Samples/Jacobi.Vst.Samples.Delay/PluginPersistence.cs
namespace Jacobi.Vst.Samples.Delay
{
    using Jacobi.Vst.Framework;
    using Jacobi.Vst.Framework.Plugin;
    using Jacobi.Vst.Framework.Plugin.IO;
    using System.IO;
    using System.Text;


    /// <summary>
    /// This class manages custom persistence for all the plugin's programs and parameters.
    /// </summary>
    internal sealed class PluginPersistence : VstPluginPersistenceBase
    {
        private readonly Plugin _plugin;

        /// <summary>
        /// Constructs a new instance.
        /// </summary>
        /// <param name="plugin">Must not be null.</param>
        public PluginPersistence(Plugin plugin)
        {
            _plugin = plugin;
        }

        protected override VstProgramReaderBase CreateProgramReader(Stream input)
        {
            return new DelayProgramReader(_plugin, input, Encoding);
        }

        private sealed class DelayProgramReader : VstProgramReaderBase
        {
            private readonly Plugin _plugin;

            public DelayProgramReader(Plugin plugin, Stream input, Encoding encoding)
                : base(input, encoding)
            {
                _plugin = plugin;
            }

            protected override VstProgram CreateProgram()
            {
                VstProgram program = new VstProgram(_plugin.ParameterFactory.Categories);

                _plugin.ParameterFactory.CreateParameters(program.Parameters);

                return program;
            }
        }
    }
}
